plasmapy/diagnostics/proton_radiography.py

```diff
@@ -300,6 +300,205 @@ def _log(self, msg):
+    ):
+        """
+        Add a wire mesh grid between the particle source and the object grid
+        that blocks particles which intersect the wires.
+
+        Parameters
+        ----------
+        location : `~astropy.units.Quantity`, shape (3)
+            A vector pointing from the origin of the grid to the center of the
+            mesh grid. This location must be between the source and the
+            object grid.
+
+            This vector will be interpreted as
+            being in either cartesian, cylindrical, or spherical coordinates
+            based on its units. Valid geometries are:
+
+            * Cartesian (x,y,z) : (meters, meters, meters)
+            * cylindrical (r, theta, z) : (meters, radians, meters)
+            * spherical (r, theta, phi) : (meters, radians, radians)
+
+            In spherical coordinates theta is the polar angle.
+
+        extent : Tuple of 1 or 2 `~astropy.units.Quantity`
+            The size of the mesh grid (in the mesh plane). If one value
+            is provided, the mesh is circular and the value provided is
+            interpreted as the diameter. If two values are provided, the
+            mesh is rectangular and they the values are interpreted as the
+            width and height respectively.
+
+        nwires : Tuple of 1 or 2 ints
+            The number of wires in the horizontal and vertical directions. If
+            only one value is provided, the number in the two directions is
+            assumed to be equal. Note that a wire will cross the center of the
+            grid only when nwires is odd.
+
+        wire_diameter : `~astropy.units.Quantity`
+            The diameter of the wires.
+
+        mesh_hdir : `numpy.ndarray`, shape (3), optional
+            A unit vector (in Cartesian coordinates) defining the horizontal
+            direction on the mesh plane. Modifying this vector can rotate the
+            mesh in the plane or tilt the mesh plane relative to the
+            source-detector axis. By default, `mesh_hdir` is set equal to
+            `detector_hdir` (see `detector_hdir` keyword in `__init__`).
+
+        mesh_vdir : `numpy.ndarray`, shape (3), optional
+            A unit vector (in Cartesian coordinates) defining the vertical
+            direction on the mesh plane. Modifying this vector can tilt the
+            mesh relative to the source-detector axis. By default, `mesh_vdir`
+            is defined to be perpendicular to `mesh_hdir` and the detector
+            plane normal (such that the mesh is parallel to the detector plane).
+
+        Raises
+        ------
+        ValueError
+            Raises a ValueError if the provided mesh location is not
+            between the source and the object grid.
+
+            Raises a ValueError if the specified grid removes ALL of the
+            particles.
+
+
+        Returns
+        -------
+
+        nremoved : int
+            The number of particles removed
```

Comment:
Okay, that's non-intuitive to me (I haven't read the notebook yet, so assume I'm a user who stumbled upon the docstring before the example).

I glanced at the test and you create the simulation, create the particles, add the wire mesh, then run the simulation. How does the wire mesh know how many particles it's going to remove before you run the simulation?

Reply:
So the wire mesh function works by calculating where particles WILL be in the future (in the wire mesh plane), determining which ones will intersect a wire, then deleting those particles before you begun `run()`. Since the mesh has to be before the grid, the system is entirely linear at that point. 

I thought about having `add_wire_mesh()` store this information in a dict, then having `run()` apply the mesh (or meshes) at the beginning of the `run()` call, but I was worried that that obscured what was happening under the hood too much.